### PR TITLE
lvs: additional support for lvs live grow and bdev resize

### PR DESCRIPTION
**Pull request by dsavitskiy**
Added helpers for support live lvs grow in client apps. Added experimental support for malloc bdev resize, which is needed for live grow tests.

### CODE REVIEW

File: include/spdk/bdev_module.h [hunk 1]
```diff
@@ -1349,6 +1349,32 @@ uint64_t spdk_bdev_io_get_submit_tsc(struct spdk_bdev_io *bdev_io);
...
+ * Change number of blocks for provided block device.
+ * It can only be called on a registered bdev.
+ *
+ * \param bdev Block device to change.
+ * \param size New size of bdev.
+ * \param resize_cb Called after the new size is applied.
+ * \param cb_arg Argument to pass to callback function.
+ * \return 0 on success, negated errno on failure.
+ */
+int spdk_bdev_resize(struct spdk_bdev *bdev,
```
Comment (tiagolobocastro) [inline include/spdk/bdev_module.h:1373]: When is this called and by whom?

File: include/spdk/lvol.h [hunk 2]
```diff
@@ -190,6 +190,45 @@ int spdk_lvs_unload(struct spdk_lvol_store *lvol_store,
 int spdk_lvs_destroy(struct spdk_lvol_store *lvol_store,
 		     spdk_lvs_op_complete cb_fn, void *cb_arg);
 
+/**
+ * Parameters for lvs initialization.
+ */
+struct spdk_lvol_opts {
```
Comment (tiagolobocastro) [inline include/spdk/lvol.h:196]: :+1:

File: include/spdk/bdev_module.h (hunk 1 above)
Comment (dsharma-dc) [reply include/spdk/bdev_module.h:1373]: We should be able to use existing bdev_event_callback for this event handling purpose I guess?

Comment (dsavitskiy) [reply include/spdk/bdev_module.h:1373]: The existing resize-related function, "spdk_bdev_notify_blockcnt_change", notifies open descriptors about Bdev size change.  "spdk_bdev_notify_blockcnt_change" locks bdev internally, to make all notifications atomic.
That works fine for Bdevs classes that do not have any internals that need to be updated during resize process. In other words, those Bdevs who have media that resizes "on its own", e.g. like a network LUN; and SPDK Bdev just keeps track on LUN size and notifies clients (open descriptors).

However, if a Bdev class needs to update its inner representation (like Malloc which has to resize its buffer), this old "spdk_bdev_notify_blockcnt_change" won't work: inner representation update and open descriptors notification will not be atomic. To make it atomic, we have to lock, update the inner representation, notify, and unlock.

So I added this "spdk_bdev_resize" that allows to pass a callback to call after locking but before notifying.

"spdk_bdev_notify_blockcnt_change" is now implemented on top of "spdk_bdev_resize".

Comment (dsavitskiy) [reply include/spdk/bdev_module.h:1373]: "spdk_bdev_resize" is called by a resize function for a Bdev, like "resize_malloc_disk".
(It happened that SPDK does not provide such resize API functions for most Bdev classes.)

Comment (dsavitskiy) [reply include/spdk/bdev_module.h:1373]: Please note, although my malloc resize implementation is done atomically in respect to client notification, it is still very unsafe in respect to I/Os .

Comment (dsharma-dc) [reply include/spdk/bdev_module.h:1373]: `spdk_bdev_notify_blockcnt_change` only updates and invokes notifications when there are no open descriptors for the bdev, isn't it?

Comment (tiagolobocastro) [reply include/spdk/bdev_module.h:1373]: > Please note, although my malloc resize implementation is done atomically in respect to client notification, it is still very unsafe in respect to I/Os .

You could use range lock api to ensure it's safe?

Comment (dsharma-dc) [reply include/spdk/bdev_module.h:1373]: I have a doubt -
1.  what IO safety we have to be concerned with here? Metadata IOs that are internally taken care by spdk should get taken care by spdk grow api. And the user/application IO should be still well within existing blobs until the blobs are expanded to pick pool's new clusters.
2. Is blob/lvol expansion and lvs expansion serialized or can be potentially racey?

Comment (dsavitskiy) [reply include/spdk/bdev_module.h:1373]: 1. It's specific to malloc: we cannot re-alloc the malloc's buffer without stopping the I/O in one way or another.
2. LVS expansion is not racey: expansion only affects LVS metadata; existing Lvols are not touched.

Comment (dsavitskiy) [reply include/spdk/bdev_module.h:1373]: > > Please note, although my malloc resize implementation is done atomically in respect to client notification, it is still very unsafe in respect to I/Os .
> 
> You could use range lock api to ensure it's safe?

Probably yes, but I guess, it is out of scope of this task.

**Review by tiagolobocastro — APPROVED** [review]

**Pull request merged**